### ISSUE TITLE
plugin-taskbar doesn't consider custom icons set in desktop entry files

### ISSUE THREAD

**Issue by pmattern** [lxde/lxqt#1262]
Custom icons set in applications' desktop entry files aren't considered by plugin-taskbar.

To reproduce just tweak key `Icon` in some desktop entry file, e. g. `Icon=thunderbird` in `firefox.desktop` and start a new LXQt session. When the application in question is launched, the Firefox icon will still be showed in plugin-taskbar.
It doesn't matter whether the icons are referred by their path or just the file name in terms of the [Icon Theme Specification](https://www.freedesktop.org/wiki/Specifications/icon-theme-spec/) like `thunderbird` above or whether it's the system file in `$XDG_DATA_DIRS` or a user-specific file in `$XDG_DATA_HOME`.

Tweaking icons this way seems a bit questionable to me as genuine application icons (acquired how?) are displayed by window managers in windows' upper left corner anyway. But nevertheless the various LXQt components should behave consistently and custom icons in desktop entry files are very well considered in plugin-mainmenu or when dragged into plugin-quicklaunch.

Yesterday's master checkouts of all LXQt comonents including lxde/lxqt-panel@94a534b.

**Comment by palinek** [comment]
Plugin-taskbar does't use .desktop files anyhow. The icon is based on what is being provided by application (and the X). Either the provided pixmap is used or the icon based on WM_CLASS.
AFAIK there is no method for connecting the running window application and a .desktop file that it was launched from.

**Comment by pmattern** [comment]
The problem exists under LXDE and Xfce as well but cannot be seen in KDE Plasma.
In Plasma icons to launch applications seem to be implemented as part of their Task Manager, similarly as proposed in #131. At least sort of as an option "Keep launchers separate" exists but does not seem to have any effect. Just saying as I was wondering whether this is related to the problem we're discussing here.
Cannot tell whether the approach of KDE is something that makes sense under LXQt, too.

**Comment by palinek** [comment]
Does KDE show an icon from .desktop file when some app is launched from e.g. terminal by running it's executable?
If the answer is yes, which .desktop file is chosen? (as there may be many .desktop files containing the very same "Exec" key)

**Comment by pmattern** [comment]
Under Plasma a custom icon set by key `Icon` of an application's desktop entry file is considered when the application is launched by invoking its binary from a terminal emulator, too.
The desktop entry file to choose the icon from seems to be selected according to the [Base Directory Specification](https://specifications.freedesktop.org/basedir-spec/latest/index.html) the same way the files are usually selected e. g. by panel menus. That is, when only a single modified desktop entry file exists in `$XDG_DATA_DIRS/applications`, the icon set in this file is selected. But when an additional modified desktop entry file exists in `$XDG_DATA_HOME/applications`, it has priority.

**Comment by tsujan** [comment]
I think the current behavior is correct; taskbar isn't about desktop files but applications themselves.

Suppose that we've made two desktop files for Firefox but with different icons. Of course, the two files are shown with different icons by FM (pcmanfm-qt, for example) but once they're clicked, Firefox will start and taskbar could only know that. The Firefox process doesn't depend on how it's started (from menu, a desktop file, or a symlink) and taskbar automatically reflects that.

**Comment by pmattern** [comment]
Assigning different icons to a particular application by different desktop entry files makes hardly ever any sense, IMO. Anyway it is not the use case this issue is referring to.
The actual problem is users are facing an inconsistency when they want to change the icon used by a particular application altogether, e. g. as they don't like what's used by default. It's the kind of problem you can read about in this [Manjaro Forum thread](https://forum.manjaro.org/t/lxqt-different-icons-for-the-same-app-in-the-quicklaunch-and-the-task-manager/16176).
As said above, a real consistency cannot be achieved anyway as all window managers are using an icon that's independent from any desktop entry file in windows' upper left corner. But if a behaviour similar to the one of Plasma's Task Manager could be implemented in plugin-taskbar with reasonable effort I still think this would be beneficial.

**Comment by tsujan** [comment]
> The actual problem is users are facing an inconsistency when they want to change the icon used by a particular application

But how? There seems to be only 2 ways for that: (1) changing the icon theme; or (2) changing the icon of the desktop file. Taskbar works with (1) but (2) is about what I tried to explain above: taskbar has nothing to do with desktop files but handles apps windows.

**Comment by palinek** [comment]
I agree with @tsujan here. Plugin taskbar should not care about searching for .desktop file to get an icon. This would then involve also watching for the .desktop files changes etc...

There is the option in plugin-taskbar to use the symbolic icon name provided by WM_CLASS and in that way the user can create/use icon theme of her/his whish to achieve the same icon as in particular .desktop file.

**Comment by pmattern** [comment]
> There is the option in plugin-taskbar to use the symbolic icon name provided by WM_CLASS and in that way the user can create/use icon theme of her/his whish to achieve the same icon as in particular .desktop file.

I hadn't thought about this before and it sounds like a very good solution to me.
If I got this right, users would have to tweak two settings, the desktop entry file and the icon associated with WM_CLASS, but that way they could accomplish complete consistency as the icon would not only be the same in plugin-mainmenu, plugin-quicklaunch and plugin-taskbar but in the upper left corner of windows as well.
Also, additional monitoring wouldn't be needed. I absolutely agree that it makes sense to avoid additional resource usage just to cover a corner case which the problem we're discussing here rather is.
There's one question remaining, though: is some reasonable documentation available how to tweak that setting? Personally I'm not aware of anything.

**Comment by palinek** [comment]
> is some reasonable documentation available how to tweak that setting?

What setting?

**Comment by pmattern** [comment]
> What setting?

I mean the steps which have to be taken to tweak the icon that is displayed via WM_CLASS.

**Comment by palinek** [comment]
> I mean the steps which have to be taken to tweak the icon that is displayed via WM_CLASS.

I don't think there is any. Basically you need to know the WM_CLASS that the application is using (e.g. by using `xprop`) and use icon theme (maybe create new one with some decent parent) that provides icon for that symbolic name.

**Comment by tsujan** [comment]
This is all about "pinning", not about changing app icons -- changing app icons should be done by changing/modifying the icon theme.

KDE Plasma can "pin" its taskbar items and change them to shortcuts. Therefore, it needs to be consistent and always use desktop entry icons. LXQt task manager (taskbar) doesn't have such a feature, so it doesn't need to consult desktop entries.
 
People come from KDE to LXQt, taking some KDE properties for granted, and expecting to see them in LXQt too. That's understandable and has helped us make LXQt better. In this special case, I think until "pinning" is not implemented in LXQt's taskbar, there's no reason to scan desktop files.

In short, and IMO, the question is whether we want "pinning". Although I don't like it, it's an important feature.

**Comment by stefonarch** [comment]
> In short, and IMO, the question is whether we want "pinning". Although I don't like it, it's an important feature.

No idea how easy or not that would be, but to have an option right click on item > "add to quicklaunch", if quicklaunch is not present it will be added to the panel.

**Comment by tsujan** [comment]
> an option right click on item > "add to quicklaunch", if quicklaunch is not present it will be added to the panel.

If quicklaunch isn't present, that will be "pinning" in the taskbar, in which case, taskbar should respect icons defined in desktop entries. I have no idea yet how easy or hard the implementation would be but it isn't something we can overlook.

**Comment by blurymind** [comment]
kde and xfce let you pin shortcuts to both the start menu and the taskbar. I am sorely missing this feature as lxqt requires the user to repeatedly search for the same 3-4 apps over and over in the menu

**Comment by tsujan** [comment]
> as lxqt requires the user to repeatedly search for the same 3-4 apps over and over in the menu

Not true. Quick Launch exists -- anyone can add it to his/her panel -- and it renders "pinning" redundant.
